plugin.go

```diff
@@ -52,16 +69,60 @@ func New(_ context.Context, next http.Handler, config *Config, name string) (htt
 }
 
 func (r *signatureVerifier) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	// Сначала пытаемся верифицировать запрос с использованием текущего ключа.
 	if err := r.verifier.VerifyRequest(req); err != nil {
 		if !errors.Is(err, ErrSignatureNotPresent) {
-			log.Printf("signature verification failed: %v", err)
+			log.Printf("signature verification failed with current key: %v", err)
 		}
 
-		if !r.cfg.DryRun {
-			http.Error(w, r.cfg.ResponseMessage, r.cfg.ResponseCode)
+		// Если верификация неудачна, пытаемся обновить ключ и повторить проверку.
+		if err := r.RetrieveAndUpdateKey(req); err != nil {
+			log.Printf("Error retrieving or updating key: %v", err)
+			http.Error(w, "Internal server error", http.StatusUnauthorized)
 			return
 		}
-	}
 
+	}
 	r.next.ServeHTTP(w, req)
 }
+
+func (cfg *Config) initSecretKey(redisClient *redis.Client) {
+	// Проверка, что URI для Redis предоставлен
+	if cfg.RedisURI == "" {
+		log.Println("REDIS_URI environment variable not set or empty")
+	}
+	ctx := context.Background()
+	key, err := redisClient.Get(ctx, "query-signature-key").Result()
+	if err != nil || key == "" {
+		log.Printf("Failed to retrieve initial key from Redis or key is empty: %v, using default key.", err)
+	} else {
+		cfg.SecretKey = key
+	}
+}
+
+func (r *signatureVerifier) RetrieveAndUpdateKey(req *http.Request) error {
```

Comment:
All logic is now very scattered. I suggest you encapsulate key management into a separate entity, e.g.:
```go

type KeyStore interface {
   Current() string
   Rorate() (string, error)
}
```
And then have 2 adapters:
```golang
type Static struct { key string }
func (s Static} Current() { return s.key }

type Redis struct { ... } 
```
It is an interesting question how to have finite/inifite rotation and when to rotate (by rotation I mean switching from old to new key), but I think it is doable.

Reply:
First, we initialize the current key in the Redis structure. Upon request, we attempt to verify it. If verification fails, we rotate the key. If the key is stale, we update it with a new one. If the current key is new but fails verification, we authorize using the old key without updating the current one.

Reply:
Just to validate I understand the logic:

Let's say we have 2 keys: correct new one (K1) and incorrect old one (K2).

1. Server receives a request with an *incorrect* signature
2. We verify signature, it fails, so we initiate key rotation by calling `Rotate`
3. `currentKey` == `K1`, so we move forward
4. `currentKey` != `K2`, so we assume it is the right one, we return it
5. We try validating a request, that still fails (since it wasn't signed either by `K1` nor `K2`, so we stop.

If I am right this probably  work. Though is not quite straighforward. I had something simpler in mind:
```golang
for _, k := r.keyStore.Keys() {
    if r.verifier.ValudateSignature(req, k) {
        return nil
   }
}

return ErrInvalidSignature
```
So the code keeps validating against all the keys there are. But redis store has a background job which rotates keys based on signals in receives from redis:
```golang
func (r *Redis) pollForChanges(ctx context.Context) {
	pubsub := r.client.PSubscribe(ctx, "__keyspace@0__:secret-key", "__keyspace@0__:secret-key-old")
	defer pubsub.Close()

	for msg := range pubsub.Channel() {
		key := msg.Channel[len("__keyspace@0__:"):]
                r.m.Lock()
                r.keys[key] = r.fetchKey(key)
                r.m.Unlock()
	}
}

func (r *Redis) Init() {
    for _, k := []string{"secret-key", "secret-key-old"} {
           key, err := r.fetchKey(k)
          if err != nil { ... }
           r.keys[k] =  key
    }
}

func (r *Redis) fetchKey() (string, error) {
         val, err := rdb.Get(ctx, key).Result()
         if err == redis.Nil {
              return "", ErrKeyDoesNotExist
        }
        if err != nil {
             return "", fmt.Errorf("unable to get key: %w", err)
       }

      return val, nil
}
```
Not sure if this is a better, but that what I had in mind. 

Also it is interesting if Yaegi spins up the whole thing per request. Then the whole thing is redundant.